### PR TITLE
Simplify TraceEnum_ELBO and add more tests

### PR DESCRIPTION
**Pull request by fritzo**
Addresses #847 

- Refactors `iter_discrete_traces()` to use `poutine.queue()` again. This appears to cost on 3% overhead on tests.
- Adds an RSVI test of `TraceEnum_ELBO` as requested in #894 . This also simplifies the test to allow use of `@pytest.mark.parametrize`.

### CODE REVIEW

File: tests/infer/test_inference.py [hunk 1]
```diff
@@ -218,59 +218,51 @@ def guide():
...
+    def guide():
+        alpha_q_log = pyro.param("alpha_q_log", torch.tensor(log_alpha_n.data + 0.17, requires_grad=True))
+        beta_q_log = pyro.param("beta_q_log", torch.tensor(log_beta_n.data - 0.143, requires_grad=True))
+        alpha_q, beta_q = torch.exp(alpha_q_log), torch.exp(beta_q_log)
+        pyro.sample("lambda_latent", gamma_dist(alpha_q, beta_q))
+
+    adam = optim.Adam({"lr": .0003, "betas": (0.97, 0.999)})
+    svi = SVI(model, guide, adam, loss="ELBO",
+              trace_graph=(elbo_impl == "TraceGraph"),
+              enum_discrete=(elbo_impl == "TraceEnum"),
```
Comment (fritzo) [inline tests/infer/test_inference.py]: Note that there's nothing discrete here; we're simply testing that DiCE correctly handles RSVI's score_parts.

Comment (martinjankowiak) [reply tests/infer/test_inference.py]: is it not worth having a KL gradient test that mixes rsvi with discretes?

Comment (fritzo) [reply tests/infer/test_inference.py]: Sure, I'll add some more test with rsvi + discrete enumeration

Comment (eb8680) [reply tests/infer/test_inference.py]: Do you want this merged now or after you add the tests?
